src/core/toga/widgets/table.py

```diff
@@ -25,127 +20,160 @@ def __repr__(self):
         return node
 
     def remove(self, node):
         self._data.remove(node)
-        if self.interface:
-            self.interface._impl.remove_node(node)
+        self._refresh()
+
+    def item(self, row: int, column: int):
+        if isinstance(row and column, int):
+            return self.data[row].data[column]
+
+    def row(self, row: int) -> TableRow:
+        if row >= 0:
+            return self.data[row]
 
+    def rows(self) -> list:
+        return self.data
 
 
 class Table(Widget):
-    """ A Table Widget allows the disply of data in the from of columns and rows.
+    """ A Table Widget allows the display of data in the from of columns and rows.
 
     Args:
         headings (``list`` of ``str``): The list of headings for the table.
         id (str): An identifier for this widget.
+        data (``list`` of ``tuple``): The data to be displayed on the table.
         style (:class:`colosseum.CSSNode`): An optional style object.
-            If no style is provided then a new one will be created for the widget.
+            If no style is provided` then a new one will be created for the widget.
+        on_select(``callable``): A function to be invoked on selecting a row of the table.
         factory (:obj:`module`): A python module that is capable to return a
             implementation of this class with the same name. (optional & normally not needed)
+
+    Examples:
+        >>> headings = ['Head 1', 'Head 2', 'Head 3']
+        >>> data = [('item 1', 'item 2', 'item3'),
+        >>>         ('item 1', 'item 2', 'item3')]
+        >>>
+        >>> table = Table(headings, data=data)
     """
+
     def __init__(self, headings, id=None, style=None, data=None, on_select=None, factory=None):
         super().__init__(id=id, style=style, factory=factory)
         self.headings = headings
-        self._data = data
+        self._data = None
         self._impl = self.factory.Table(interface=self)
+        self.data = data
 
         self.on_select = on_select
 
     @property
     def data(self):
-        '''
-        :returns: The data source of the tree
-        :rtype: ``dict``
-        '''
-        return self._data
+        """ The data source of the widget. It accepts table data
+        in the form of ``list``, ``tuple``, or :obj:`ListDataSource`
+
+        Returns:
+            (list) Returns a ``list`` of lists. Where the outer lists represents the
+            rows and each inner list represents a column.
+        """
+        return self._data if self._data is not None else None
 
     @data.setter
     def data(self, data):
-        '''
-        Set the data source of the data
-
-        :param data: Data source
-        :type  data: ``list``, ``tuple``, or ``class``
-        '''
         if isinstance(data, (list, tuple)):
-            self._data = ListDataSource(data)
+            self._data = ListDataSource(data, refresh_function=self._impl.refresh)
         else:
             self._data = data
```

Comment:
Given that we claim a promise above about what we accept, it would be nice to throw an error here if `data` is not an iterable or a ListDataSource.

...

Thinking about this more, the only promise necessary is that `self._data` is some kind of iterable, right? I mean, we could (theoretically) pass in a generator as the data source, couldn't we?

Could we check for that here?

Reply:
I disagree that this is needed. One of the use-cases for this is to pass in an arbitrary object that adheres to the ListDataSource interface, but isn't actually backed by a ListDataSource. As long as the data source quacks the same, we don't care what object it is.